keps/sig-architecture/3136-beta-apis-off-by-default/README.md

```diff
@@ -0,0 +1,304 @@
+When moving from beta to GA, it will be important for sigs to explicitly seek feedback.
+
+If beta APIs are off by default, it is possible that sigs don't treat taking an API as an indication of a "mostly-baked" API.
+If this happens, then more transformation may be required.
+Keeping our beta API rules consistent and continuing to enforce easy to use APIs seems to be the best option.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### Test Plan
+
+Integration tests will be written to ensure that no new beta APIs are enabled in the kube-apiserver by default.
+Unit tests will be written to ensure that the new flag functionality works as expected.
+
+### Graduation Criteria
+
+This KEP is a policy KEP, not a feature KEP.  It will start as GA.
+
+#### GA
+
+- Integration and unit tests from above.
+- updating the enablement docs for beta
+  - https://kubernetes.io/docs/reference/using-api/#api-versioning
+  - https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/#using-a-feature 
+    Even though that is talking about feature gates, it is likely worth calling out there that new beta REST APIs are no
+    longer enabled by default)
+- email to dev@kubernetes.io to explain the new policy
+- blog post explaining change in time for 1.24 release
+- CI configuration updated to have a testing mode that enables beta APIs, likely set using `kube-apiserver --runtime-config=api/beta=true`
+
+### Upgrade / Downgrade Strategy
+
+The additional command line flag format for `--runtime-config` will not be recognized on older levels of kubernetes.
+This means that when downgrading, cluster-admins will have to adjust their CLI arguments if they opted into a new beta API.
+This is congruent to flag handling for new features today.
+Because this only impacts new beta APIs, there is no behavior change for existing APIs on upgrade.
+
+### Version Skew Strategy
+
+Because this only impacts new beta APIs, there is no novel skew risk.
+
+## Production Readiness Review Questionnaire
+
+Not applicable because this is a policy KEP.
+
+## Implementation History
+
+<!--
+Major milestones in the lifecycle of a KEP should be tracked in this section.
+Major milestones might include:
+- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
+- the `Proposal` section being merged, signaling agreement on a proposed design
+- the date implementation started
+- the first Kubernetes release where an initial version of the KEP was available
+- the version of Kubernetes where the KEP graduated to general availability
+- when the KEP was retired or superseded
+-->
+
+## Drawbacks
+
+<!--
+Why should this KEP _not_ be implemented?
+-->
+
+## Alternatives
```

Comment:
An alternative - or maybe more accurately a follow up - could be something like the "API lenses" idea @thockin proposed a couple years ago here: https://docs.google.com/document/d/1bjyu7ksEM2sNpIyA625zj8XHt_5xtaX_z2EtvA7_fLQ/edit#heading=h.5m3e1uuqb7dd

That proposal is mostly talking about adding/modifying *fields* within an API. However, a similar concept can be applied to whole APIs. For example, this could amount to a "hidden" setting for beta APIs: the API is enabled, but in order to use it, you need to add a special HTTP header. This would avoid the inadvertent "just copy and paste this manifest from the internet" type of spread of APIs. That is, our current method of designated use of a beta API - just slightly different value in a yaml file - is too easy and accessible. The "special header" (or some other mechanism) makes it a more explicit choice (at least until some big provider just sets up all their tooling with that enabled always...).

Reply:
> The "special header" (or some other mechanism) makes it a more explicit choice (at least until some big provider just sets up all their tooling with that enabled always...).

that's exactly what would happen :-/

opt-in things like that are super hard to propagate through pipelines that compose/wrap other tools or manifests, so things designed to handle generic manifests opt into every possible option to work with as many sources as possible.

Reply:
The main point here is that this provides a mechanism that:
1) Avoids inadvertent reliance on beta features
2) Allows enablement of beta features on a cluster while still requiring the end user to consciously "opt in" to the use of that beta feature.

It's somewhat of a middle ground, recognizing that in a shared cluster environment, a specific team may want to rely on a beta feature, but we do not want to expose other users of that cluster to inadvertent risk of creating tech debt.

My proposal would be to implement this KEP - i.e., stop the bleeding/creation of new debt for our users - and then follow up as quickly as possible with something like this that shifts the beta API consumption decision to end users that are well-aware of what they are doing (and only if allowed by the admin, of course).

Reply:
@liggitt, sorry didn't see your comment before I made my second comment. It may be that that would happen. But that would be the providers choice, not the projects choice. That argument also can be used to say "this KEP is useless, people will just explicitly set them to 'on' always".

As I see it, as a project have a couple responsibilities:
1) To our users, we should make it easier to do things the *right* way rather than the wrong way. Today (beta by default), at the current level of scale and maturity, we have this balance wrong. This KEP addresses that.
2) To our feature developers, to ensure we give them the tools they need to build the best features they can, and make it as easy as possible for them to validate that with their users.

I do think the we can do more for 2). I think some proposal like this may be useful in that regard. But that's subject to debate. There are also other ways too help. For example, we can build (via enhancements process or contribex, perhaps), programs to make it easy to recruit and collect feedback from beta testers - a stable set of companies / teams / individuals that have said they are willing to sign up for betas, perhaps. And tooling to provide feedback channels that users can access to review beta features. Even clusters we can stand up, enabled for beta, for them to use (beware the crypto miner hackers :( ). 

For those folks that are worried about a loss in feedback for features - it would be good to start to put together some ideas in this area.

Reply:
yes, good points. however some folks are concerned about the amount of feedback we'll get. rather than have every SIG re-invent the mechanisms, we may be able to help. but I agree it's not clear any help is needed.

Reply:
> I'm assuming that every feature we're building (especially when we're talking about full APIs - because that's what we're talking here), we are not building "in the air".

That's fair, but a lot of the time, probably specifically for API machinery, we create features that will enable people to build "stuff", and they don't know they can build these "stuff" until they try it out. We're building features for way more people than our mere customer base, and we see companies thrive by using our features in creative ways.

Also related to that, a lot of the time, teams can't really rely on a feature until it's enabled on all supported cluster versions. The release process being what it is, that's already a very long time. Turning beta features off will only make that step longer (and more risky). I think this change generally increases the GTM time and risk, two metrics that people are generally trying to decrease.